src/netif.c

```diff
@@ -2603,40 +2680,24 @@ static void kni_ingress(struct rte_mbuf *mbuf, struct netif_port *dev,
     /* VLAN device cannot be scheduled by kni_send2kern_loop */
     if ((dev->type == PORT_TYPE_VLAN && qconf->kni_len > 0) ||
         unlikely(qconf->kni_len == NETIF_MAX_PKT_BURST)) {
-        rte_spinlock_lock(&kni_lock);
-        pkt_num = rte_kni_tx_burst(dev->kni.kni, qconf->kni_mbufs, qconf->kni_len);
-        rte_spinlock_unlock(&kni_lock);
-
-        if (unlikely(pkt_num < qconf->kni_len)) {
-            RTE_LOG(WARNING, NETIF, "%s: fail to send pkts to kni\n", __func__);
-            free_mbufs(&(qconf->kni_mbufs[pkt_num]), qconf->kni_len - pkt_num);
-        }
-
-        qconf->kni_len = 0;
+            kni_ingress_pkts_redirect(dev, qconf);
```

Comment:
Vlan device uses physical device's qconf, so vlan may not need be processed in particular. Can you test it?